packages/vm/src/runTx.ts

```diff
@@ -281,11 +279,11 @@ async function _runTx(this: VM, opts: RunTxOpts): Promise<RunTxResult> {
         `sender doesn't have enough funds to send tx. The upfront cost is: ${cost} and the sender's account only has: ${balance}`
       )
     }
-    if (tx instanceof FeeMarketEIP1559Transaction) {
+    if (tx.supportsEIP(1559)) {
       // EIP-1559 spec:
       // The signer must be able to afford the transaction
       // `assert balance >= gas_limit * max_fee_per_gas`
-      const cost = tx.gasLimit.mul(tx.maxFeePerGas)
+      const cost = tx.gasLimit.mul((tx as FeeMarketEIP1559Transaction).maxFeePerGas)
```

Comment:
It might be interesting to consider turning the `supportsEIP` method into a TypeScript typeguard. This would allow  automatic inference of the transaction type (no `as FeeMarketEIP1559Transaction` would then be necessary).

Reply:
Oh that's a really cool idea, could you commit it to this branch or sketch out how we may do that? I am just reading about it, super interesting: https://www.typescriptlang.org/docs/handbook/advanced-types.html#using-type-predicates

Reply:
The idea is really cool yeah! 😄 It doesn't apply on this context though I would very much say. The `supportsEIP()` (or however we call it method) is explicitly not about the tx type but about some certain capability. So for 2930 (access lists) it is just not decidable on what to cast (lowest common denominator `AccessListEIP2930Transaction` or `FeeMarketEIP1559Transaction`). I would assume it is rather not good to make assumptions here (e.g.: let's cast to the lowest common denominator) since this might not be appropriate for the specific usage context. And at the end there is no such thing as "transaction hierarchy", tx types are at the end completely independent entities, so it might even be wrong to talk about "lowest common denominator" in this context at all.

And with a potential EIP-155 addition this concept breaks completely.

Alternative ideas:

Might it be possible to either:

a) add this to the `type` getter instead, this then *gets* deterministic on this call and this would also be more consistent from a usage context perspective I guess (if someone *wants* to assume an 1559 tx in e.g. an `if` clause one can also use a `tx.type === 2` switch)
b) if possible likely optimal: add this typeguard after the constructor is called (no idea if this is technically possible, also never have no insights on typeguards)

> An additional thought: could we also refactor the _signedTxImplementsEIP155 (packages\tx\src\legacyTransaction.ts) boolean to use that supportsEIP method also?

Yes, I will add this along the rework, makes super-much sense, txs working in a EIP-155 context is a super-important tx capability and we are dealing with this in a far too implicit manner for my taste right now. I guess we need to go the way here to deprecate `_signedTxImplementsEIP155()` methods - even if considered private - since there is some higher chance that people are calling this specific one for various custom reasons.